src/webnn/native/dml/DMLUtils.cpp

```diff
@@ -52,42 +69,395 @@ namespace webnn::native::dml {
+        std::vector<ArrayBufferView> outputArrayBufferViews;
+        ArrayBufferView outputArrayBufferView;
+        auto outputNodes = compiledGraph->outputNodes;
+        for (size_t i = 0; i < outputNodes.size(); ++i) {
+            std::string name = outputNodes[i].name;
+            auto namedOutputs = outputs->GetRecords();
+            if (namedOutputs.find(name) != namedOutputs.end()) {
+                outputArrayBufferView = namedOutputs[name].arrayBufferView;
+                outputArrayBufferViews.push_back(outputArrayBufferView);
+                DAWN_ASSERT(outputArrayBufferView.buffer != nullptr &&
+                            outputArrayBufferView.byteLength != 0);
+            } else {
+                size_t byteLength = reinterpret_cast<const DML_BUFFER_TENSOR_DESC*>(
+                                        outputNodes[i].outputTensorDesc.Desc)
+                                        ->TotalTensorSizeInBytes;
+                // It is an unuseful output of dml graph. We need not read back and copy buffer
+                // to it, just reserve it as a placeholder.
+                outputArrayBufferView = {nullptr, byteLength, 0};
+                outputArrayBufferViews.push_back(outputArrayBufferView);
+            }
+        }
+
+        std::vector<DML_BINDING_DESC> outputBindingDesc(outputNodes.size());
+        std::vector<DML_BUFFER_BINDING> outputBufferBinding(outputNodes.size());
+
+        uint64_t outputOffset = 0;
+        for (size_t i = 0; i < outputNodes.size(); ++i) {
+            auto output = outputArrayBufferViews[i];
+            outputOffset = RoundUpToMultiple(
+                outputOffset, static_cast<uint64_t>(DML_MINIMUM_BUFFER_TENSOR_ALIGNMENT));
+            outputBufferBinding[i].Buffer = mOutputResource.Get();
+            outputBufferBinding[i].Offset = outputOffset;
+            outputBufferBinding[i].SizeInBytes = output.byteLength;
+            outputBindingDesc[i] = {DML_BINDING_TYPE_BUFFER, &outputBufferBinding[i]};
+            outputOffset = outputOffset + output.byteLength;
+        }
+        mBindingTable->BindOutputs(outputBindingDesc.size(), outputBindingDesc.data());
+
+        // Record execution of the compiled operator.
+        ID3D12DescriptorHeap* descriptorHeaps[] = {mDescriptorHeap.Get()};
+        mCommandList->SetDescriptorHeaps(ARRAYSIZE(descriptorHeaps), descriptorHeaps);
+        mCommandRecorder->RecordDispatch(mCommandList.Get(), compiledGraph->compiledOperator.Get(),
+                                         mBindingTable.Get());
+
+        // Copy buffer from outputResource to readBackResource.
+        CopyBufferRegion(mOutputResource, mReadBackResource, mOutputResourceSize,
+                         D3D12_RESOURCE_STATE_COPY_SOURCE, false);
+        CloseExecuteResetWait();
+
+        int8_t* readBackBuffer;
+        WEBNN_CHECK(mReadBackResource->Map(0, nullptr, reinterpret_cast<void**>(&readBackBuffer)));
```

Comment:
What happens if CopyBufferRegion() or ExecuteCommandLists doesn't complete before memcpy starts? 

Remember, these calls will return immediately where the GPU executes them like a background thread. Unless you "tell" the CPU to wait for a result, it wont.

Reply:
Will the `fence->SetEventOnCompletion` in ` CloseExecuteResetWait()` wait until the fence is completed?

Reply:
No, it will not. That will "signal" the fence (or set the fence-value). 

If you want the CPU to wait for the fence, use `WaitForSingleObject`. If you want the GPU to wait for the fence, use `ID3D12Queue::Wait`.

Reply:
Just like the usage here https://github.com/microsoft/DirectML/blob/master/Samples/HelloDirectML/main.cpp#L98?